tests/lib/rules/define-macros-order.js

```diff
@@ -0,0 +1,389 @@
+      options: optionsEmitsFirst,
+      errors: [
+        {
+          message: message('defineEmits'),
+          line: 12
+        }
+      ]
+    },
+    {
+      filename: 'test.vue',
+      code: `
+        <script lang="ts" setup>
+          import bla from 'bla';
+          interface Foo {};
+          type Bar = {};
+          // <--- auto-fix should move \`defineProps\` here
+          const someOtherCode = '';
+          import foo from 'bar'; // not idiomatic, but allowed
+          interface SomeOtherInterface {};
+          defineProps({ test: Boolean });
+        </script>
+      `,
+      output: `
+        <script lang="ts" setup>
+          import bla from 'bla';
+          interface Foo {};
+          type Bar = {};
+          // <--- auto-fix should move \`defineProps\` here
+          defineProps({ test: Boolean });
+          const someOtherCode = '';
+          import foo from 'bar'; // not idiomatic, but allowed
+          interface SomeOtherInterface {};
+        </script>
+      `,
+      parserOptions: {
+        parser: require.resolve('@typescript-eslint/parser')
+      },
+      errors: [
+        {
+          message: message('defineProps'),
+          line: 10
+        }
+      ]
+    },
+    {
+      filename: 'test.vue',
+      code: `
+        <script lang="ts" setup>
+          debugger
+
+          console.log('test1')
+
+          /** Description for props */
+          const props = withDefaults(defineProps<Props>(), {
+            msg: 'hello'
+          });
+
+          console.log('test2')
+
+          // Description for emit
+          // Description for emit line 2
+          const emit = defineEmits<{(e: 'test'): void}>();
+
+          console.log('test3')
+        </script>
+      `,
+      output: `
+        <script lang="ts" setup>
+          debugger
+
+          // Description for emit
+          // Description for emit line 2
+          const emit = defineEmits<{(e: 'test'): void}>();
+          /** Description for props */
+          const props = withDefaults(defineProps<Props>(), {
+            msg: 'hello'
+          });
+          console.log('test1')
+
+
+          console.log('test2')
+
+
+          console.log('test3')
+        </script>
+      `,
+      parserOptions: {
+        parser: require.resolve('@typescript-eslint/parser')
+      },
+      options: optionsEmitsFirst,
+      errors: [
+        {
+          message: message('defineEmits'),
+          line: 16
+        }
+      ]
+    },
+    {
+      filename: 'test.vue',
+      code: `
+        <script setup>
+          const props = defineProps({ test: Boolean });defineEmits(['update:test'])
+        </script>
+      `,
+      output: `
+        <script setup>
+defineEmits(['update:test'])
+          const props = defineProps({ test: Boolean });        </script>
+      `,
+      options: optionsEmitsFirst,
+      errors: [
+        {
+          message: message('defineEmits'),
+          line: 3
+        }
+      ]
+    }
```

Comment:
@ota-meshi @FloEdelmann 
There is the moment I need help.
Need to decide what we "move".
![image](https://user-images.githubusercontent.com/41228762/164217601-e5f05187-1b4c-4c63-9aee-df4cd1ddb5f9.png)

1. We can move only the node itself(without newlines).
This breaks the code formatting almost at every case
```vue
<-- BEFORE -->
<script setup>
    defineEmits(['update:test']);
    const props = defineProps({ test: Boolean });
    console.log('test')
</script>

<-- AFTER -->
<script setup>
    const props = defineProps({ test: Boolean });defineEmits(['update:test']);

    console.log('test')
</script>
```

2. We can move node with space before and space with a new line after
In most cases, this behavior doesn't break formatting drastically but does make some double new lines
```vue
<-- BEFORE -->
<script setup>
    defineEmits(['update:test'])

    console.log('test')

    const props = defineProps({ test: Boolean })

    console.log('test2')
</script>

<-- AFTER -->
<script setup>
    const props = defineProps({ test: Boolean })
    defineEmits(['update:test'])
    
    console.log('test')


    console.log('test2')
</script>
```

3. Cut node with newlines before\after
This thing has the same downsides as 2nd

I took the 2nd variant. I.e. there would be cases where the result code contains multiple new lines

There is an option to make some heuristics about the newlines, but it feels that this is not the right direction

Reply:
How about moving from the start position of the node to the position immediately before the next token?

```vue
<script setup>
    defineEmits(['update:test'])

    console.log('test')

    |<const props = defineProps({ test: Boolean })

    >|console.log('test2')
</script>
```


```vue
<script setup>
    |<const props = defineProps({ test: Boolean })

    >|defineEmits(['update:test'])

    console.log('test')

    console.log('test2')
</script>
```

Reply:
There would be problem with this case
```vue
<script setup>
    defineEmits(['update:test'])

    console.log('test')

    |<const props = defineProps({ test: Boolean })
>|</script>
```

Here is the expected behavior:
https://user-images.githubusercontent.com/41228762/164386145-98ae4695-29a8-4023-9f63-e792141165c5.mp4
It seems that it's too hard to move every node one by one...

It feels that we need smth like this:
1. Cut node without newlines
2. Remove all empty lines after the node
3. Paste the node on the top with the same new lines that the top node have

I'll try to write it. Sounds not so fearful...

Reply:
Oh. I didn't consider that case.

I think it might work if you cut from immediately after the previous token to the end of the node, and move it immediately after the token before the insert node. What do you think?
It may look a bit out of format because it looks like more line breaks at the beginning, but I don't think it's a problem for the scope of auto-fix.
I want to avoid leaving new multiple blank lines as much as possible.

```vue
<script setup>|target
    defineEmits(['update:test'])

    console.log('test')|<

    const props = defineProps({ test: Boolean })>|
</script>

<script setup>|<

    const props = defineProps({ test: Boolean })>|
    defineEmits(['update:test'])

    console.log('test')
</script>
```

```vue
<script setup>|target
    defineEmits(['update:test'])

    console.log('test')|<

    const props = defineProps({ test: Boolean })>|

    console.log('test2')
</script>

<script setup>|<

    const props = defineProps({ test: Boolean })>|
    defineEmits(['update:test'])

    console.log('test')

    console.log('test2')
</script>
```

Reply:
> Cut node without newlines
Remove all empty lines after the node
Paste the node on the top with the same new lines that the top node have

Oh. I now understand this. It may work.